server/server.go

```diff
@@ -524,3 +526,60 @@ type httpErrorResponse struct {
 	Error            string `json:"error"`
 	ErrorDescription string `json:"error_description,omitempty"`
 }
+
+// isFunnelRequest checks if the request is coming through Tailscale Funnel
+func isFunnelRequest(r *http.Request) bool {
+	// If we're funneling through the local tailscaled, it will set this HTTP header
+	if r.Header.Get("Tailscale-Funnel-Request") != "" {
+		return true
+	}
+
+	// If the funneled connection is from tsnet, then the net.Conn will be of type ipn.FunnelConn
+	netConn := r.Context().Value(CtxConn{})
+	// if the conn is wrapped inside TLS, unwrap it
+	if tlsConn, ok := netConn.(*tls.Conn); ok {
+		netConn = tlsConn.NetConn()
+	}
+	if _, ok := netConn.(*ipn.FunnelConn); ok {
+		return true
+	}
+	return false
+}
+
+// csrfCheck wraps the next handler in a CSRF check
+func (s *IDPServer) csrfCheck(next http.HandlerFunc) http.HandlerFunc {
```

Comment:
Implemented this here instead of using `http.NewCrossOriginProtection` because we're stuck on go v1.24.x until the govulncheck panic is fixed: https://github.com/golang/go/issues/73871